manual/qpdf.1
.\"
.\" This file is automatically generated by generate_auto_job.
.\" Edits will be automatically overwritten if the build is
.\" run in maintainer mode.
.\"
.TH QPDF "1" "" "qpdf version 12.4.2" "User Commands"
.SH NAME
qpdf \- PDF transformation software
.SH SYNOPSIS
.B qpdf
.RI "[ " options " ] " infilename " [ " outfilename " ]"
.SH DESCRIPTION
The qpdf program is used to convert one PDF file to another equivalent
PDF file.  It is capable of performing a variety of transformations
such as linearization (also known as web optimization or fast web
viewing), encryption, and decryption of PDF files.  It also has many
options for inspecting or checking PDF files, some of which are
useful primarily to PDF developers.
.PP
For a summary of qpdf's options, please run \fBqpdf \-\-help\fR. A
complete manual can be found at https://qpdf.readthedocs.io.
.PP
In some distributions, you can find offline versions of the qpdf
manual in /usr/share/doc/qpdf. This may require installing an
additional package, such as \fBqpdf-doc\fR in Debian-based Linux
distributions. You can also download offline documentation from
https://qpdf.readthedocs.io. For more information, see README-doc.txt
in the source distribution.
.SH USAGE (basic invocation)
Read a PDF file, apply transformations or modifications, and write
a new PDF file.

Usage: qpdf [infile] [options] [outfile]
   OR  qpdf --help[={topic|--option}]

.IP \[bu]
infile, options, and outfile may be in any order as long as infile
precedes outfile.
.IP \[bu]
Use --empty in place of an input file for a zero-page, empty input
.IP \[bu]
Use --replace-input in place of an output file to overwrite the
input file with the output
.IP \[bu]
outfile may be - to write to stdout; reading from stdin is not supported
.IP \[bu]
@filename is an argument file; each line is treated as a separate
command-line argument
.IP \[bu]
@- may be used to read arguments from stdin
.IP \[bu]
Later options may override earlier options if contradictory
.PP
Related Options:
.TP
.B --empty \-\- use empty file as input
Use in place of infile for an empty input. Especially useful
with --pages.
.TP
.B --replace-input \-\- overwrite input with output
Use in place of outfile to overwrite the input file with the output.
.TP
.B --job-json-file \-\- job JSON file
--job-json-file=file

Specify the name of a file whose contents are expected to
contain a QPDFJob JSON file. Run qpdf --job-json-help for a
description of the JSON input file format.
.SH EXIT-STATUS (meanings of qpdf's exit codes)
Meaning of exit codes:

.IP \[bu]
0: no errors or warnings
.IP \[bu]
1: not used by qpdf but may be used by the shell if unable to invoke qpdf
.IP \[bu]
2: errors detected
.IP \[bu]
3: warnings detected, unless --warning-exit-0 is given
.PP
Related Options:
.TP
.B --warning-exit-0 \-\- exit 0 even with warnings
Use exit status 0 instead of 3 when warnings are present. When
combined with --no-warn, warnings are completely ignored.
.SH COMPLETION (shell completion)
Shell completion is supported with bash and zsh. Use
eval "$(qpdf --completion-bash)" or eval "$(qpdf --completion-zsh)"
to enable. The QPDF_EXECUTABLE environment variable overrides the
path to qpdf that these commands output.
.PP
Related Options:
.TP
.B --completion-bash \-\- enable bash completion
Output a command that enables bash completion
.TP
.B --completion-zsh \-\- enable zsh completion
Output a command that enables zsh completion
.SH HELP (information about qpdf)
Help options provide some information about qpdf itself. Help
options are only valid as the first and only command-line argument.
.PP
Related Options:
.TP
.B --help \-\- provide help
--help[=--option|topic]

--help: provide general information and a list of topics
--help=--option: provide help on a specific option
--help=topic: provide help on a topic
.TP
.B --version \-\- show qpdf version
Display the version of qpdf.
.TP
.B --copyright \-\- show copyright information
Display copyright and license information.
.TP
.B --show-crypto \-\- show available crypto providers
Show a list of available crypto providers, one per line. The
default provider is shown first.
.TP
.B --job-json-help \-\- show format of job JSON
Describe the format of the QPDFJob JSON input used by
--job-json-file.
.TP
.B --zopfli \-\- indicate whether zopfli is enabled and active
If zopfli support is compiled in, indicate whether it is active,
and exit normally. Otherwise, indicate that it is not compiled
in, and exit with an error code. If zopfli is compiled in,
activate it by setting the ``QPDF_ZOPFLI`` environment variable.
.SH GENERAL (general options)
General options control qpdf's behavior in ways that are not
directly related to the operation it is performing.
.PP
Related Options:
.TP
.B --password \-\- password for encrypted file
--password=password

Specify a password for an encrypted, password-protected file.
Not needed for encrypted files without a password.
.TP
.B --password-file \-\- read password from a file
--password-file=filename

The first line of the specified file is used as the password.
This is used in place of the --password option.
.TP
.B --verbose \-\- print additional information
Output additional information about various things qpdf is
doing, including information about files created and operations
performed.
.TP
.B --progress \-\- show progress when writing
Indicate progress when writing files.
.TP
.B --no-warn \-\- suppress printing of warning messages
Suppress printing of warning messages. If warnings were
encountered, qpdf still exits with exit status 3.
Use --warning-exit-0 with --no-warn to completely ignore
warnings.
.TP
.B --deterministic-id \-\- generate ID deterministically
Generate a secure, random document ID only using static
information, such as the page contents. Does not use the file's
name or attributes or the current time.
.TP
.B --allow-weak-crypto \-\- allow insecure cryptographic algorithms
Allow creation of files with weak cryptographic algorithms. This
option is necessary to create 40-bit files or 128-bit files that
use RC4 encryption.
.TP
.B --keep-files-open \-\- manage keeping multiple files open
--keep-files-open=[y|n]

When qpdf needs to work with many files, as when merging large
numbers of files, explicitly indicate whether files should be
kept open. The default behavior is to determine this based on
the number of files.
.TP
.B --keep-files-open-threshold \-\- set threshold for --keep-files-open
--keep-files-open-threshold=count

Set the threshold used by --keep-files-open, overriding the
default value of 200.
.SH ADVANCED-CONTROL (tweak qpdf's behavior)
Advanced control options control qpdf's behavior in ways that would
normally never be needed by a user but that may be useful to
developers or people investigating problems with specific files.
.PP
Related Options:
.TP
.B --password-is-hex-key \-\- provide hex-encoded encryption key
Provide the underlying file encryption key as a hex-encoded
string rather than supplying a password. This is an expert
option.
.TP
.B --suppress-password-recovery \-\- don't try different password encodings
Suppress qpdf's usual behavior of attempting different encodings
of a password that contains non-ASCII Unicode characters if the
first attempt doesn't succeed.
.TP
.B --password-mode \-\- tweak how qpdf encodes passwords
--password-mode=mode

Fine-tune how qpdf controls encoding of Unicode passwords. Valid
options are auto, bytes, hex-bytes, and unicode.
.TP
.B --suppress-recovery \-\- suppress error recovery
Avoid attempting to recover when errors are found in a file's
cross reference table or stream lengths.
.TP
.B --ignore-xref-streams \-\- use xref tables rather than streams
Ignore any cross-reference streams in the file, falling back to
cross-reference tables or triggering document recovery.
.SH TRANSFORMATION (make structural PDF changes)
The options below tell qpdf to apply transformations that change
the structure without changing the content.
.PP
Related Options:
.TP
.B --linearize \-\- linearize (web-optimize) output
Create linearized (web-optimized) output files.
.TP
.B --encrypt \-\- start encryption options
--encrypt [options] --

Run qpdf --help=encryption for details.
.TP
.B --decrypt \-\- remove encryption from input file
Create an unencrypted output file even if the input file was
encrypted. Normally qpdf preserves whatever encryption was
present on the input file. This option overrides that behavior.
.TP
.B --remove-restrictions \-\- remove security restrictions from input file
Remove restrictions associated with digitally signed PDF files.
This may be combined with --decrypt to allow free editing of
previously signed/encrypted files. This option invalidates and
disables any digital signatures but leaves their visual
appearances intact.
.TP
.B --copy-encryption \-\- copy another file's encryption details
--copy-encryption=file

Copy encryption details from the specified file instead of
preserving the input file's encryption. Use --encryption-file-password
to specify the encryption file's password.
.TP
.B --encryption-file-password \-\- supply password for --copy-encryption
--encryption-file-password=password

If the file named in --copy-encryption requires a password, use
this option to supply the password.
.TP
.B --qdf \-\- enable viewing PDF code in a text editor
Create a PDF file suitable for viewing in a text editor and even
editing. This is for editing the PDF code, not the page contents.
All streams that can be uncompressed are uncompressed, and
content streams are normalized, among other changes. The
companion tool "fix-qdf" can be used to repair hand-edited QDF
files. QDF is a feature specific to the qpdf tool. Please see
the "QDF Mode" chapter in the manual.
.TP
.B --no-original-object-ids \-\- omit original object IDs in qdf
Omit comments in a QDF file indicating the object ID an object
had in the original file.
.TP
.B --compress-streams \-\- compress uncompressed streams
--compress-streams=[y|n]

Setting --compress-streams=n prevents qpdf from compressing
uncompressed streams. This can be useful if you are leaving some
streams uncompressed intentionally.
.TP
.B --decode-level \-\- control which streams to uncompress
--decode-level=parameter

When uncompressing streams, control which types of compression
schemes should be uncompressed:
.IP \[bu]
none: don't uncompress anything. This is the default with
--json-output.
.IP \[bu]
generalized: uncompress streams compressed with a
general-purpose compression algorithm. This is the default
except when --json-output is given.
.IP \[bu]
specialized: in addition to generalized, also uncompress
streams compressed with a special-purpose but non-lossy
compression scheme
.IP \[bu]
all: in addition to specialized, uncompress streams compressed
with lossy compression schemes like JPEG (DCT)
qpdf does not know how to uncompress all compression schemes.
.TP
.B --stream-data \-\- control stream compression
--stream-data=parameter

This option controls how streams are compressed in the output.
It is less granular than the newer options, --compress-streams
and --decode-level.

Parameters:
.IP \[bu]
compress: same as --compress-streams=y --decode-level=generalized
.IP \[bu]
preserve: same as --compress-streams=n --decode-level=none
.IP \[bu]
uncompress: same as --compress-streams=n --decode-level=generalized
.TP
.B --recompress-flate \-\- uncompress and recompress flate
The default generalized compression scheme used by PDF is flate,
which is the same as used by zip and gzip. Usually qpdf just
leaves these alone. This option tells qpdf to uncompress and
recompress streams compressed with flate. This can be useful
when combined with --compression-level.
.TP
.B --compression-level \-\- set compression level for flate
--compression-level=level

Set a compression level from 1 (least, fastest) to 9 (most,
slowest) when compressing files with flate (used in zip and
gzip), which is the default compression for most PDF files.
You need --recompress-flate with this option if you want to
change already compressed streams.
.TP
.B --jpeg-quality \-\- set jpeg quality level for jpeg
--jpeg-quality=level

When rewriting images with --optimize-images, set a quality
level from 0 (lowest) to 100 (highest) for writing new images.
Higher quality results in larger images, and lower quality
results in smaller images. This option is only effective when
combined with --optimize-images.
.TP
.B --normalize-content \-\- fix newlines in content streams
--normalize-content=[y|n]

Normalize newlines to UNIX-style newlines in PDF content
streams, which is useful for viewing them in a programmer's text
editor across multiple platforms. This is also turned on by
--qdf.
.TP
.B --object-streams \-\- control use of object streams
--object-streams=mode

Control what qpdf does regarding object streams. Options:
.IP \[bu]
preserve: preserve original object streams, if any (the default)
.IP \[bu]
disable: create output files with no object streams
.IP \[bu]
generate: create object streams, and compress objects when possible
.TP
.B --preserve-unreferenced \-\- preserve unreferenced objects
Preserve all objects from the input even if not referenced.
.TP
.B --remove-unreferenced-resources \-\- remove unreferenced page resources
--remove-unreferenced-resources=parameter

Remove from a page's resource dictionary any resources that are
not referenced in the page's contents. Parameters: "auto"
(default), "yes", "no".
.TP
.B --preserve-unreferenced-resources \-\- use --remove-unreferenced-resources=no
Synonym for --remove-unreferenced-resources=no. Use that instead.
.TP
.B --newline-before-endstream \-\- force a newline before endstream
For an extra newline before endstream. Using this option enables
qpdf to preserve PDF/A when rewriting such files.
.TP
.B --coalesce-contents \-\- combine content streams
If a page has an array of content streams, concatenate them into
a single content stream.
.TP
.B --externalize-inline-images \-\- convert inline to regular images
Convert inline images to regular images.
.TP
.B --ii-min-bytes \-\- set minimum size for --externalize-inline-images
--ii-min-bytes=size-in-bytes

Don't externalize inline images smaller than this size. The
default is 1,024. Use 0 for no minimum.
.TP
.B --min-version \-\- set minimum PDF version
--min-version=version

Force the PDF version of the output to be at least the specified
version. The version number format is
"major.minor[.extension-level]", which sets the version header
to "major.minor" and the extension level, if specified, to
"extension-level".
.TP
.B --force-version \-\- set output PDF version
--force-version=version

Force the output PDF file's PDF version header to be the specified
value, even if the file uses features that may not be available
in that version.
.SH PAGE-RANGES (page range syntax)
A full description of the page range syntax, with examples, can be
found in the manual. In summary, a range is a comma-separated list
of groups. A group is a number or a range of numbers separated by a
dash. A group may be prepended by x to exclude its members from the
previous group. A number may be one of

.IP \[bu]
<n>        where <n> represents a number is the <n>th page
.IP \[bu]
r<n>       is the <n>th page from the end
.IP \[bu]
z          the last page, same as r1

.IP \[bu]
a,b,c      pages a, b, and c
.IP \[bu]
a-b        pages a through b inclusive; if a > b, this counts down
.IP \[bu]
a-b,xc     pages a through b except page c
.IP \[bu]
a-b,xc-d   pages a through b except pages c through d

You can append :even or :odd to select every other page from the
resulting set of pages, where :odd starts with the first page and
:even starts with the second page. These are odd and even pages
from the resulting set, not based on the original page numbers.
.SH MODIFICATION (change parts of the PDF)
Modification options make systematic changes to certain parts of
the PDF, causing the PDF to render differently from the original.
.PP
Related Options:
.TP
.B --pages \-\- begin page selection
--pages [--file=]file [options] [...] --

Run qpdf --help=page-selection for details.
.TP
.B --file \-\- source for pages
--file=file

Specify the file for the current page operation. This is used
with --pages, --overlay, and --underlay and appears between the
option and the terminating --. Run qpdf --help=page-selection
for details.
.TP
.B --range \-\- page range
--range=numeric-range

Specify the page range for the current page operation with
--pages. If omitted, all pages are selected. This is used
with --pages and appears between --pages and --. Run
qpdf --help=page-selection for details.
.TP
.B --collate \-\- collate with --pages
--collate[=n[,m,...]]

Collate rather than concatenate pages specified with --pages.
With a numeric parameter, collate in groups of n. The default
is 1. With comma-separated numeric parameters, take n from the
first file, m from the second, etc. Run
qpdf --help=page-selection for additional details.
.TP
.B --split-pages \-\- write pages to separate files
--split-pages[=n]

This option causes qpdf to create separate output files for each
page or group of pages rather than a single output file.

File names are generated from the specified output file as follows:

.IP \[bu]
If the string %d appears in the output file name, it is replaced with a
zero-padded page range starting from 1
.IP \[bu]
Otherwise, if the output file name ends in .pdf (case insensitive), a
zero-padded page range, preceded by a dash, is inserted before the file
extension
.IP \[bu]
Otherwise, the file name is appended with a zero-padded page range
preceded by a dash.

Page ranges are single page numbers for single-page groups or first-last
for multi-page groups.
.TP
.B --overlay \-\- begin overlay options
--overlay file [options] --

Overlay pages from another PDF file on the output.
Run qpdf --help=overlay-underlay for details.
.TP
.B --underlay \-\- begin underlay options
--underlay file [options] --

Underlay pages from another PDF file on the output.
Run qpdf --help=overlay-underlay for details.
.TP
.B --flatten-rotation \-\- remove rotation from page dictionary
For each page that is rotated using the /Rotate key in the
page's dictionary, remove the /Rotate key and implement the
identical rotation semantics by modifying the page's contents.
This can be useful if a broken PDF viewer fails to properly
consider page rotation metadata.
.TP
.B --flatten-annotations \-\- push annotations into content
--flatten-annotations=parameter

Push page annotations into the content streams. This may be
necessary in some case when printing or splitting files.
Parameters: "all", "print", "screen".
.TP
.B --rotate \-\- rotate pages
--rotate=[+|-]angle[:page-range]

Rotate specified pages by multiples of 90 degrees specifying
either absolute or relative angles. "angle" may be 0, 90, 180,
or 270. You almost always want to use +angle or -angle rather
than just angle, as discussed in the manual. Run
qpdf --help=page-ranges for help with page ranges.
.TP
.B --generate-appearances \-\- generate appearances for form fields
PDF form fields consist of values and appearances, which may be
inconsistent with each other if a form field value has been
modified without updating its appearance. This option tells qpdf
to generate new appearance streams. There are some limitations,
which are discussed in the manual.
.TP
.B --optimize-images \-\- use efficient compression for images
Attempt to use DCT (JPEG) compression for images that fall
within certain constraints as long as doing so decreases the
size in bytes of the image. See also help for the following
options:
  --oi-min-width
  --oi-min-height
  --oi-min-area
  --keep-inline-images
.TP
.B --oi-min-width \-\- minimum width for --optimize-images
--oi-min-width=width

Don't optimize images whose width is below the specified value.
.TP
.B --oi-min-height \-\- minimum height for --optimize-images
--oi-min-height=height

Don't optimize images whose height is below the specified value.
.TP
.B --oi-min-area \-\- minimum area for --optimize-images
--oi-min-area=area-in-pixels

Don't optimize images whose area in pixels is below the specified value.
.TP
.B --keep-inline-images \-\- exclude inline images from optimization
Prevent inline images from being considered by --optimize-images.
.TP
.B --remove-acroform \-\- remove the interactive form dictionary
Exclude the interactive form dictionary from the output file. This
option only removes the interactive form dictionary from the
document catalog. It does not remove form field dictionaries or
widget annotations. See the manual for more detail.
.TP
.B --remove-info \-\- remove file information
Exclude file information (except modification date) from the output file.
.TP
.B --remove-metadata \-\- remove metadata
Exclude metadata from the output file.
.TP
.B --remove-page-labels \-\- remove explicit page numbers
Exclude page labels (explicit page numbers) from the output file.
.TP
.B --remove-structure \-\- remove metadata
Exclude the structure tree from the output file.
.TP
.B --set-page-labels \-\- number pages for the entire document
--set-page-labels label-spec ... --

Set page labels (explicit page numbers) for the entire file.
Each label-spec has the form

first-page:[type][/start[/prefix]]

where

.IP \[bu]
"first-page" represents a sequential page number using the
same format as page ranges: a number, a number preceded by "r"
to indicate counting from the end, or "z" indicating the last
page
.IP \[bu]
"type" is one of
- D: Arabic numerals (digits)
- A: Upper-case alphabetic characters
- a: Lower-case alphabetic characters
- R: Upper-case Roman numerals
- r: Lower-case Roman numerals
- omitted: the page number does not appear, though the prefix,
  if specified will still appear
.IP \[bu]
"start" must be a number >= 1
.IP \[bu]
"prefix"` may be any string and is prepended to each page
label

The first label spec must have a first-page value of 1,
indicating the first page of the document. If multiple page
label specs are specified, they must be given in increasing
order.

If multiple page label specs are specified, they must be given
in increasing order.

A given page label spec causes pages to be numbered according to
that scheme starting with first-page and continuing until the
next label spec or the end of the document. If you want to omit
numbering starting at a certain page, you can use first-page: as
the spec.

Example: "1:r 5:D" would number the first four pages i through
iv, then the remaining pages with Arabic numerals starting with
1 and continuing sequentially until the end of the document. For
additional examples, please consult the manual.
.SH ENCRYPTION (create encrypted files)
Create encrypted files. Usage:

--encrypt \
  [--user-password=user-password] \
  [--owner-password=owner-password] \
  --bits=key-length [options] --

OR

--encrypt user-password owner-password key-length [options] --

The first form, with flags for the passwords and bit length, was
introduced in qpdf 11.7.0. Only the --bits option is mandatory.
This form allows you to use any text as the password. If passwords
are specified, they must be given before the --bits option.

The second form has been in qpdf since the beginning and will
continue to be supported. Either or both of user-password and
owner-password may be empty strings.

The key-length parameter must be either 40, 128, or 256. The user
and/or owner password may be omitted. Omitting either password
enables the PDF file to be opened without a password. Specifying
the same value for the user and owner password and specifying an
empty owner password are both considered insecure.

Encryption options are terminated by "--" by itself.

40-bit encryption is insecure, as is 128-bit encryption without
AES. Use 256-bit encryption unless you have a specific reason to
use an insecure format, such as testing or compatibility with very
old viewers. You must use the --allow-weak-crypto flag to create
encrypted files that use insecure cryptographic algorithms. The
--allow-weak-crypto flag appears outside of --encrypt ... --
(before --encrypt or after --).

Available options vary by key length. Not all readers respect all
restrictions. Different PDF readers respond differently to various
combinations of options. Sometimes a PDF viewer may show you
restrictions that differ from what you selected. This is probably
not a bug in qpdf.

Options for 40-bit only:
  --annotate=[y|n]         restrict comments, filling forms, and signing
  --extract=[y|n]          restrict text/graphic extraction
  --modify=[y|n]           restrict document modification
  --print=[y|n]            restrict printing

Options for 128-bit or 256-bit:
  --accessibility=[y|n]    restrict accessibility (usually ignored)
  --annotate=[y|n]         restrict commenting/filling form fields
  --assemble=[y|n]         restrict document assembly
  --extract=[y|n]          restrict text/graphic extraction
  --form=[y|n]             restrict filling form fields
  --modify-other=[y|n]     restrict other modifications
  --modify=modify-opt      control modify access by level
  --print=print-opt        control printing access
  --cleartext-metadata     prevent encryption of metadata

For 128-bit only:
  --use-aes=[y|n]          indicates whether to use AES encryption
  --force-V4               forces use of V=4 encryption handler

For 256-bit only:
  --force-R5               forces use of deprecated R=5 encryption
  --allow-insecure         allow user password with empty owner password

Values for print-opt:
  none                     disallow printing
  low                      allow only low-resolution printing
  full                     allow full printing

Values for modify-opt:
  none                     allow no modifications
  assembly                 allow document assembly only
  form                     assembly + filling in form fields and signing
  annotate                 form + commenting and modifying forms
  all                      allow full document modification
.PP
Related Options:
.TP

ChangeLog
2025-02-19  Jay Berkenbilt  <ejb@ql.org>

        * END USE OF ChangeLog. From this point forward, please consult
        the release notes for important changes and version control
        history for detailed changes.

2025-02-15  Jay Berkenbilt  <ejb@ql.org>

	* 11.10.1: release

	* Detect cygwin as Windows for fix-qdf tests that don't work on
	Windows. The perl used for the test runner in GitHub Actions
	changed. This was causing some non-Windows tests to fail and also
	causing path canonicalization to fail for absolute paths with
	drive letters.

	* Detect zopfli by looking for <zopfli/zopfli.h> rather than just
	<zopfli.h>

2025-02-08  Jay Berkenbilt  <ejb@ql.org>

	* 11.10.0: release

2025-02-02  Jay Berkenbilt  <ejb@ql.org>

	* Add support for the zopfli compression library. See manual for
	details. Fixes #1323.

	* Have fix-qdf accept a second argument, interpreted as the output
	file. Fixes #1330.

2024-02-01  M Holger  <m.holger@qpdf.org>

	* Bug fix: in qpdf CLI / QPDFJob throw a QPDFUsage exception if a
	parameter is given for a flag. Previously the parameter was ignored and
	e.g. --encrypt=n was treated as --encrypt. Fixes #1329.

2024-09-20  Chao Li  <mslichao@outlook.com>

	* Add C API function qpdf_oh_free_buffer to release memory allocated
	by stream data functions.

2024-09-19  M Holger  <m.holger@qpdf.org>

	* Bug fix: QPDFWriter stream DecodeLevel incorrectly defaulted to
	none instead of generalized. Fixes #1286.

2024-08-25  M Holger  <m.holger@qpdf.org>

	* Add new command-line arguments --remove-metadata and --remove-info
	to exclude document metadata and information from the output PDF
	file. Partially fixes #1145.

2024-08-06  M Holger  <m.holger@qpdf.org>

	* Bug fix: when writing real numbers as JSON ensure that they don't
	have a trailing decimal point. Fixes #1261.

2024-07-14  M Holger  <m.holger@qpdf.org>

	* Bug fix: handle named destinations where the entry is a
	dictionary with /D entry instead of an explicit destination.
	Fixes #1238.

2024-07-04  M Holger  <m.holger@qpdf.org>

	* Treat corrupt JPEG streams as unfilterable. This avoids them
	getting uncompressed when writing PDF files with decode level all.

2024-07-02  Jay Berkenbilt  <ejb@ql.org>

	* Add QPDF::setMaxWarnings to set the maximum of warnings before
	warning suppression.

	* Add static option to Pl_DCT to limit memory usage of
	decompression. The option is generally exposed but is primarily
	intended to support fuzz tests, which have explicit memory limits
	that are smaller than what is commonly seen in the wild with PDF
	files.

	* Add static option to Pl_DCT to control whether decompression of
	corrupt JPEG data is attempted.

2024-07-01  M Holger  <m.holger@qpdf.org>

	* Bug fix: certain invalid object streams caused the insertion of
	invalid entries into in the xref table.

2024-06-29  M Holger  <m.holger@qpdf.org>

	* Bug fix: in QPDFOutlineObjectHelper detect loops in the list of
	direct children of an outline item.

2024-06-27  M Holger  <m.holger@qpdf.org>

	* Add sanity check in QPDF xref table reconstruction to reject
	objects with impossibly large object id in order to improve
	handling of severely damaged PDF files.

2024-06-25  M Holger  <m.holger@qpdf.org>

	* Detect severely damaged PDF files early. After parsing the xref
	table in QPDF throw a damagedPDF exception if the root of the pages
	tree is not a dictionary.

2024-06-07  Jay Berkenbilt  <ejb@ql.org>

	* 11.9.1: release

2024-06-06  Jay Berkenbilt  <ejb@ql.org>

	* Rewrite a recursive function to be iterative in linearization to
	enable more complex files to be linearized, especially on Windows.

2024-02-24  Jay Berkenbilt  <ejb@ql.org>

	* Avoid non-standard use of `std::basic_string_view`. Contribution
	from Zoe Clifford.

2024-02-24  Jay Berkenbilt  <ejb@ql.org>

	* 11.9.0: release

2024-02-17  Jay Berkenbilt  <ejb@ql.org>

	* Add ENABLE_COVERAGE cmake option to assist with generating
	coverage reports.

	* From M. Holger: add QPDFObjectHandle::writeJSON to directly
	write a JSON representation to a pipeline. This is much faster
	than writing the serialized result of getJSON.

2024-02-11  Jay Berkenbilt  <ejb@ql.org>

	* The previous fix to #1056 was incomplete. When setting a check
	box value, the previous fix allowed any value other than /Off to
	mean checked. Now we also set the actual value based on the
	allowable non-/Off value in the normal appearance dictionary.
	Fixes #1056.

2024-02-03  Jay Berkenbilt  <ejb@ql.org>

	* Add fuzz testing for JSON.

2024-01-15  Jay Berkenbilt  <ejb@ql.org>

	* Add JSON::getDictItem (from m-holger)

2024-01-10  Jay Berkenbilt  <ejb@ql.org>

	* Allow --overlay and --underlay to be repeated. They may appear
	multiple times on the command-line and will be stacked in the
	order in which they appear. In QPDFJob JSON, the overlay and
	underlay keys may contain arrays. For compatibility, they may also
	contain a single dictionary.

2024-01-09  Jay Berkenbilt  <ejb@ql.org>

	* Add new command-line arguments --file and --range which can be
	used within --pages in place of positional arguments. Allow --file
	to be used inside of --overlay and --underlay as well. These new
	options can be freely intermixed with positional arguments. Also
	add file(), range(), and password() to QPDFJob::PagesConfig as an
	alternative to pageSpec.

2024-01-08  Jay Berkenbilt  <ejb@ql.org>

	* 11.8.0: release

2024-01-07  Jay Berkenbilt  <ejb@ql.org>

	* Bug fix: treat references to older generations of objects as
	null.

2024-01-06  Jay Berkenbilt  <ejb@ql.org>

	* When recovering a file's xref table, attempt to find xref
	streams if a traditional trailer dictionary is not found. Fixes
	#1103.

2024-01-05  Jay Berkenbilt  <ejb@ql.org>

	* Add --set-page-labels command-line argument and supporting API.
	Fixes #939.
	  - QPDFJob::Config::setPageLabels
	  - pdf_page_label_e enumerated type
	  - QPDFPageLabelDocumentHelper::pageLabelDict

2024-01-01  Jay Berkenbilt  <ejb@ql.org>


manual/_ext/qpdf.py
from collections import defaultdict
from operator import itemgetter
import re

from sphinx import addnodes
from sphinx.directives import ObjectDescription
from sphinx.domains import Domain, Index
from sphinx.roles import XRefRole
from sphinx.util.nodes import make_refnode

# Reference:
#   https://www.sphinx-doc.org/en/master/development/tutorials/todo.html
#   https://www.sphinx-doc.org/en/master/development/tutorials/recipe.html

# cSpell:ignore contnode
# cSpell:ignore docname
# cSpell:ignore docnames
# cSpell:ignore localname
# cSpell:ignore refnode
# cSpell:ignore signode


class OptionDirective(ObjectDescription):
    has_content = True

    def handle_signature(self, sig, signode):
        signode += addnodes.desc_name(text=sig)
        return sig

    def add_target_and_index(self, name_cls, sig, signode):
        m = re.match(r'^--([^\[= ]+)', sig)
        if not m:
            raise Exception('option must start with --')
        option_name = m.group(1)
        signode['ids'].append(f'option-{option_name}')
        qpdf = self.env.get_domain('qpdf')
        qpdf.add_option(sig, option_name)


class OptionIndex(Index):
    name = 'options'
    localname = 'qpdf Command-line Options'
    shortname = 'Options'

    def generate(self, docnames=None):
        content = defaultdict(list)
        options = self.domain.get_objects()
        options = sorted(options, key=itemgetter(0))

        # name, subtype, docname, anchor, extra, qualifier, description
        for name, display_name, typ, docname, anchor, _ in options:
            m = re.match(r'^(--([^\[= ]+))', display_name)
            if not m:
                raise Exception(
                    'OptionIndex.generate: display name not as expected')
            content[m.group(2)[0].lower()].append(
                (m.group(1), 0, docname, anchor, '', '', typ))

        content = sorted(content.items())
        return content, True


class QpdfDomain(Domain):
    name = 'qpdf'
    label = 'qpdf documentation domain'
    roles = {
        'ref': XRefRole()
    }
    directives = {
        'option': OptionDirective,
    }
    indices = {
        OptionIndex,
    }
    initial_data = {
        'options': [],  # object list
    }

    def get_full_qualified_name(self, node):
        return '{}.{}'.format('option', node.arguments[0])

    def get_objects(self):
        for obj in self.data['options']:
            yield(obj)

    def resolve_xref(self, env, from_doc_name, builder, typ, target, node,
                     contnode):
        match = [(docname, anchor)
                 for name, sig, typ, docname, anchor, priority
                 in self.get_objects() if name == f'option.{target[2:]}']

        if len(match) > 0:
            to_doc_name = match[0][0]
            match_target = match[0][1]
            return make_refnode(builder, from_doc_name, to_doc_name,
                                match_target, contnode, match_target)
        else:
            raise Exception(f'invalid option xref ({target})')

    def add_option(self, signature, option_name):
        if self.env.docname != 'cli':
            raise Exception(
                'qpdf:option directives don\'t work outside of cli.rst')

        name = f'option.{option_name}'
        anchor = f'option-{option_name}'

        # name, display_name, type, docname, anchor, priority
        self.data['options'].append(
            (name, signature, '', self.env.docname, anchor, 0))

    def purge_options(self, docname):
        self.data['options'] = list([
            x for x in self.data['options']
            if x[3] != docname
        ])


def purge_options(app, env, docname):
    option = env.get_domain('qpdf')
    option.purge_options(docname)


def setup(app):
    app.add_domain(QpdfDomain)
    app.connect('env-purge-doc', purge_options)

    return {
        'version': '0.1',
        'parallel_read_safe': True,
        'parallel_write_safe': True,
    }
